plugins/providers/hyperv/action/configure.rb

```diff
@@ -98,14 +153,220 @@ def call(env)
           end
 
           if env[:machine].provider_config.enable_enhanced_session_mode
-            env[:ui].detail(I18n.t("vagrant.hyperv_enable_enhanced_session"))
             env[:machine].provider.driver.set_enhanced_session_transport_type("HvSocket")
           else
-            env[:ui].detail(I18n.t("vagrant.hyperv_disable_enhanced_session"))
             env[:machine].provider.driver.set_enhanced_session_transport_type("VMBus")
           end
-
+          
+          # Continue the middleware chain.
           @app.call(env)
+
+          # If we have networks to configure, then we configure it now, since
+          # that requires the machine to be up and running.
+          if !adapters.empty? && !networks.empty?
+            assign_interface_numbers(networks, adapters)
+
+            # Only configure the networks the user requested us to configure
+            networks_to_configure = networks.select { |n| n[:auto_config] }
+            if !networks_to_configure.empty?
+              env[:ui].info I18n.t("vagrant.actions.vm.network.configuring")
+              # Hyper-v is aggressive with dhcp renewal and when the VM runs `netplan apply` it might get
+              # a new IP address and break the SSH connection which will hang forever
+              begin
+                Timeout.timeout(60) do
+                  env[:machine].guest.capability(:configure_networks, networks_to_configure)
+                end
+              rescue Timeout::Error
+                env[:ui].detail("Timed out waiting for network configuration to apply but it's probably all good")
```

Comment:
This is a bit annoying, I haven't looked at a more graceful way of handling the VM disconnecting the remote connection while it's refreshing it's network. The NAT DHCP server seems to not provide a long lease on IPs so they roll over really fast and when a VM applies network changes the IP changes causing the connection to hang.